go-controller/pkg/ovn/egressip.go

```diff
@@ -206,6 +219,165 @@ func (oc *Controller) syncEgressIPs(eIPs []interface{}) {
 			klog.Errorf("Unable to delete legacy logical router policy: %s, stderr: %s, err: %v", policyID, stderr, err)
 		}
 	}
+
+	// This part will take of syncing stale data which we might have in OVN if
+	// there's no ovnkube-master running for a while, while there are changes to
+	// pods/egress IPs.
+	// It will sync:
+	// - Egress IPs which have been deleted while ovnkube-master was down
+	// - pods/namespaces which have stopped matching on egress IPs while
+	//   ovnkube-master was down
+	if egressIPToPodIPCache, err := oc.generatePodIPCacheForEgressIP(eIPs); err == nil {
+		oc.syncStaleEgressReroutePolicy(egressIPToPodIPCache)
+		oc.syncStaleNATRules(egressIPToPodIPCache)
+	}
+}
+
+func (oc *Controller) syncStaleEgressReroutePolicy(egressIPToPodIPCache map[string]sets.String) {
+	policyItems, stderr, err := util.RunOVNNbctl(
+		"--format=csv",
+		"--data=bare",
+		"--no-heading",
+		"--columns=_uuid,external_ids,match",
+		"find",
+		"logical_router_policy",
+		fmt.Sprintf("priority=%v", types.EgressIPReroutePriority),
+	)
+	if err != nil {
+		klog.Errorf("Unable to sync egress IPs, unable to find logical router policies, stderr: %s, err: %v", stderr, err)
+		return
+	}
+	for _, policyItem := range strings.Split(policyItems, "\n") {
+		if policyItem == "" {
+			continue
+		}
+		policyFields := strings.Split(policyItem, ",")
+
+		UUID := policyFields[0]
+		externalID := policyFields[1]
+		match := policyFields[2]
+
+		// A match condition for egress IPs will look like:
+		// ip4.src == 10.244.2.5
+		splitMatch := strings.Split(match, " ")
+		logicalIP := splitMatch[len(splitMatch)-1]
+		parsedLogicalIP := net.ParseIP(logicalIP)
+		if parsedLogicalIP == nil {
+			klog.Errorf("Unable to parse logical_ip: %s from match condition: %s", logicalIP, match)
+			continue
+		}
+		egressIPName := strings.Split(externalID, "=")[1]
+
+		podIPCache, exists := egressIPToPodIPCache[egressIPName]
+		if !exists || !podIPCache.Has(parsedLogicalIP.String()) {
+			_, stderr, err := util.RunOVNNbctl(
+				"remove",
+				"logical_router",
+				types.OVNClusterRouter,
+				"policies",
+				UUID,
+			)
+			if err != nil {
+				klog.Errorf("Unable to remove stale logical router policy for EgressIP: %s, stderr: %s, err: %v", egressIPName, stderr, err)
+			}
+		}
+	}
+}
+
+func (oc *Controller) syncStaleNATRules(egressIPToPodIPCache map[string]sets.String) {
+	natItems, stderr, err := util.RunOVNNbctl(
+		"--format=csv",
+		"--data=bare",
+		"--no-heading",
+		"--columns=_uuid,external_ids,logical_ip",
+		"find",
+		"nat",
+	)
+	if err != nil {
+		klog.Errorf("Unable to sync egress IPs, unable to find NAT IDs, stderr: %s, err: %v", stderr, err)
+		return
+	}
+	for _, natItem := range strings.Split(natItems, "\n") {
+		if natItem == "" {
+			continue
+		}
+		natFields := strings.Split(natItem, ",")
```

Comment:
need to check for length of `natFields`?

Reply:
See my comment: https://github.com/ovn-org/ovn-kubernetes/pull/2300/commits/57743e3c4c99451dc96208f2f74c0a69fbde2816#r684472664